File: src/lib/engine/catalyst-events.ts
import { TICKER_BY_COINMARKETCAL_ID } from "./asset-ids";

/**
 * CoinMarketCal → catalyst_event normalization. Pure and fixture-testable:
 * the worker's context pass fetches, this file decides what qualifies.
 *
 * Filtering philosophy (high-signal over volume):
 * - Coins are mapped by PROVIDER ID via asset-ids.ts; unmapped coins drop.
 * - A credibility gate keeps only events with a proof link, enough votes, or
 *   high provider confidence — the crowd-sourced tail is noise by default.
 * - percentOfSupply stays null on the free tier: the prompt layer must then
 *   present an unlock as a scheduling fact, never a supply-pressure signal.
 */

export type CatalystKind = "unlock" | "listing" | "upgrade" | "fork" | "burn" | "other";

export interface CatalystCredibility {
  votes: number | null;
  confidencePct: number | null;
  hotScore: number | null;
}

export interface CatalystEventInput {
  symbol: string;
  kind: CatalystKind;
  title: string;
  description: string | null;
  occursAt: string;
  source: string;
  sourceId: string | null;
  url: string | null;
  credibility: CatalystCredibility | null;
  percentOfSupply: number | null;
  dedupKey: string;
}

// Category name → kind, first match wins. Checked lowercase-inclusive so
// "Token Unlock", "Unlocks" and "unlock" all land the same place.
const KIND_BY_CATEGORY: Array<[RegExp, CatalystKind]> = [
  [/unlock|vesting/, "unlock"],
  [/listing|exchange/, "listing"],
  [/fork|swap/, "fork"],
  [/burn/, "burn"],
  [/release|update|upgrade|mainnet|testnet|hard fork/, "upgrade"],
];

const MIN_VOTES = 15;
const MIN_CONFIDENCE_PCT = 80;

function asString(value: unknown): string | null {
  if (typeof value === "string" && value.trim()) return value.trim();
  // CoinMarketCal titles/descriptions arrive as { en: "..." }.
  const en = (value as { en?: unknown } | null)?.en;
  return typeof en === "string" && en.trim() ? en.trim() : null;
}

function asNumber(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function classifyKind(categories: unknown): CatalystKind {
  const names = Array.isArray(categories)
    ? categories
        .map((c) => asString((c as { name?: unknown })?.name) ?? "")
        .join(" ")
        .toLowerCase()
    : "";
  for (const [re, kind] of KIND_BY_CATEGORY) if (re.test(names)) return kind;
  return "other";
}

/** proof link OR enough votes OR high provider confidence — else it's noise. */
export function passesCredibilityGate(input: {
  url: string | null;
  votes: number | null;
  confidencePct: number | null;
}): boolean {
  if (input.url) return true;
  if (input.votes !== null && input.votes >= MIN_VOTES) return true;
  if (input.confidencePct !== null && input.confidencePct >= MIN_CONFIDENCE_PCT) return true;
  return false;
}

/**
 * Normalize one CoinMarketCal `/v1/events` response page. Events fan out one
 * row per mapped coin; a whole-payload shape surprise returns [] (the caller
 * records an ingest error when the fetch failed — an empty page is valid).
 */
export function normalizeCoinMarketCalEvents(
  payload: unknown,
  nowMs: number,
): CatalystEventInput[] {
  const body = (payload as { body?: unknown } | null)?.body;
  if (!Array.isArray(body)) return [];
  const out: CatalystEventInput[] = [];
  for (const raw of body) {
    const ev = raw as Record<string, unknown>;
    const title = asString(ev.title);
    const occursAtMs = Date.parse(asString(ev.date_event) ?? "");
    if (!title || !Number.isFinite(occursAtMs)) continue;
    // Past events (beyond a day of timezone slack) are not "what matters next".
    if (occursAtMs < nowMs - 24 * 60 * 60_000) continue;

    const sourceId = asString(ev.id) ?? (asNumber(ev.id) !== null ? String(ev.id) : null);
    const url = asString(ev.proof) ?? asString(ev.source);
    const votes = asNumber(ev.vote_count);
    const confidencePct = asNumber(ev.percentage);
    const hotScore = asNumber(ev.hot_index) ?? asNumber(ev.trending_index);
    if (!passesCredibilityGate({ url, votes, confidencePct })) continue;

    const kind = classifyKind(ev.categories);
    const coins = Array.isArray(ev.coins) ? ev.coins : [];
    for (const coin of coins) {
      const providerId = asString((coin as { id?: unknown })?.id);
      const symbol = providerId ? TICKER_BY_COINMARKETCAL_ID[providerId] : undefined;
      if (!symbol) continue; // unmapped coin — dropped by design
      out.push({
        symbol,
        kind,
        title,
        description: asString(ev.description),
        occursAt: new Date(occursAtMs).toISOString(),
        source: "coinmarketcal",
        sourceId,
        url,
        credibility: { votes, confidencePct, hotScore },
        percentOfSupply: null, // not exposed on the free tier — see header
        dedupKey: `coinmarketcal:${sourceId ?? `${title}:${occursAtMs}`}:${symbol}`,
      });
    }
  }
  return out;
}

File: engine/smc/catalyst_events.py
"""CoinMarketCal → catalyst_event normalization. Pure and fixture-testable:
the worker's context pass fetches, this file decides what qualifies.

Filtering philosophy (high-signal over volume):
- Coins are mapped by PROVIDER ID via asset_ids.py; unmapped coins drop.
- A credibility gate keeps only events with a proof link, enough votes, or
  high provider confidence — the crowd-sourced tail is noise by default.
- percent_of_supply stays None on the free tier: the prompt layer must then
  present an unlock as a scheduling fact, never a supply-pressure signal.

Port of catalyst-events.ts.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from datetime import UTC, datetime
from email.utils import parsedate_to_datetime
from typing import Literal

from smc.asset_ids import TICKER_BY_COINMARKETCAL_ID

CatalystKind = Literal["unlock", "listing", "upgrade", "fork", "burn", "other"]


@dataclass(slots=True)
class CatalystCredibility:
    votes: float | None
    confidence_pct: float | None
    hot_score: float | None


@dataclass(slots=True)
class CatalystEventInput:
    symbol: str
    kind: CatalystKind
    title: str
    description: str | None
    occurs_at: str
    source: str
    source_id: str | None
    url: str | None
    credibility: CatalystCredibility | None
    percent_of_supply: float | None
    dedup_key: str


# Category name → kind, first match wins. Checked lowercase-inclusive so
# "Token Unlock", "Unlocks" and "unlock" all land the same place.
_KIND_BY_CATEGORY: list[tuple[re.Pattern[str], CatalystKind]] = [
    (re.compile(r"unlock|vesting"), "unlock"),
    (re.compile(r"listing|exchange"), "listing"),
    (re.compile(r"fork|swap"), "fork"),
    (re.compile(r"burn"), "burn"),
    (re.compile(r"release|update|upgrade|mainnet|testnet|hard fork"), "upgrade"),
]

_MIN_VOTES = 15
_MIN_CONFIDENCE_PCT = 80


def _as_string(value: object) -> str | None:
    if isinstance(value, str) and value.strip():
        return value.strip()
    # CoinMarketCal titles/descriptions arrive as { en: "..." }.
    if isinstance(value, dict):
        en = value.get("en")
        if isinstance(en, str) and en.strip():
            return en.strip()
    return None


def _as_number(value: object) -> float | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)) and math.isfinite(value):
        return float(value)
    return None


def _classify_kind(categories: object) -> CatalystKind:
    if isinstance(categories, list):
        names = " ".join(
            _as_string(c.get("name")) or "" for c in categories if isinstance(c, dict)
        ).lower()
    else:
        names = ""
    for pattern, kind in _KIND_BY_CATEGORY:
        if pattern.search(names):
            return kind
    return "other"


def passes_credibility_gate(
    url: str | None, votes: float | None, confidence_pct: float | None
) -> bool:
    """Proof link OR enough votes OR high provider confidence — else noise."""
    if url:
        return True
    if votes is not None and votes >= _MIN_VOTES:
        return True
    return confidence_pct is not None and confidence_pct >= _MIN_CONFIDENCE_PCT


def _parse_date_ms(value: str | None) -> float:
    if not value:
        return math.nan
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp() * 1000
    except ValueError:
        pass
    try:
        return parsedate_to_datetime(value).timestamp() * 1000
    except (TypeError, ValueError):
        return math.nan


def normalize_coinmarketcal_events(payload: object, now_ms: float) -> list[CatalystEventInput]:
    """Normalize one CoinMarketCal ``/v1/events`` response page. Events fan out
    one row per mapped coin; a whole-payload shape surprise returns [] (the
    caller records an ingest error when the fetch failed — an empty page is
    valid)."""
    body = payload.get("body") if isinstance(payload, dict) else None
    if not isinstance(body, list):
        return []
    out: list[CatalystEventInput] = []
    for raw in body:
        if not isinstance(raw, dict):
            continue
        title = _as_string(raw.get("title"))
        occurs_at_ms = _parse_date_ms(_as_string(raw.get("date_event")))
        if not title or not math.isfinite(occurs_at_ms):
            continue
        # Past events (beyond a day of timezone slack) are not "what matters next".
        if occurs_at_ms < now_ms - 24 * 60 * 60_000:
            continue

        source_id = _as_string(raw.get("id"))
        if source_id is None:
            id_num = _as_number(raw.get("id"))
            source_id = str(int(id_num)) if id_num is not None else None
        url = _as_string(raw.get("proof")) or _as_string(raw.get("source"))
        votes = _as_number(raw.get("vote_count"))
        confidence_pct = _as_number(raw.get("percentage"))
        hot_score = _as_number(raw.get("hot_index"))
        if hot_score is None:
            hot_score = _as_number(raw.get("trending_index"))
        if not passes_credibility_gate(url, votes, confidence_pct):
            continue

        kind = _classify_kind(raw.get("categories"))
        coins = raw.get("coins")
        occurs_at = (
            datetime.fromtimestamp(occurs_at_ms / 1000, tz=UTC)
            .isoformat(timespec="milliseconds")
            .replace("+00:00", "Z")
        )
        for coin in coins if isinstance(coins, list) else []:
            provider_id = _as_string(coin.get("id")) if isinstance(coin, dict) else None
            symbol = TICKER_BY_COINMARKETCAL_ID.get(provider_id) if provider_id else None
            if symbol is None:
                continue  # unmapped coin — dropped by design
            article_key = source_id if source_id is not None else f"{title}:{int(occurs_at_ms)}"
            out.append(
                CatalystEventInput(
                    symbol=symbol,
                    kind=kind,
                    title=title,
                    description=_as_string(raw.get("description")),
                    occurs_at=occurs_at,
                    source="coinmarketcal",
                    source_id=source_id,
                    url=url,
                    credibility=CatalystCredibility(
                        votes=votes, confidence_pct=confidence_pct, hot_score=hot_score
                    ),
                    percent_of_supply=None,  # not exposed on the free tier
                    dedup_key=f"coinmarketcal:{article_key}:{symbol}",
                )
            )
    return out
